### DISCUSSION TITLE
陽性登録の意思をアプリから事前登録する機能

### DISCUSSION THREAD

**Discussion by parasite2006** [cocoa-mhlw/cocoa#257] [Ideas]
## その機能リクエストは何らかの問題に関連しますか / Is your feature request related to a problem?

現行の陽性登録手順では、保健所から陽性検査結果の電話連絡が来てからでないと陽性登録の意思表示ができない。しかも保健所は陽性登録の意思表示を聞いてからHER-SYSへのデータ登録を開始して本人確認用の処理番号発行操作を実行するので、検査結果通知から処理番号が届いて陽性登録を実行するまでの間に相当のタイムラグが生じてしまう。

## 解決策についてお書きください / Describe the solution you'd like

HER-SYSへのデータ登録開始時期を前倒しして検査結果を知ったユーザーができるだけ時間をおかずに陽性登録を実行できるよう、検査結果が届く前の段階で陽性登録の意思をアプリから登録できる機能を追加する。

## あなたが考える代替案についてご説明ください / Describe alternatives you've considered

アプリのホーム画面上に「検査結果が陽性なら登録を希望する」といったボタンを追加で設ける。検査を受けることが決まってから結果が連絡されるまでの間にこのボタンを押し、次の画面上で「検査結果が陽性なら登録を希望しますか？」に「はい」を選ぶと次のいずれかの反応が生じる：

1) 折り返し陽性登録に使えるバーコードまたは番号が発行される
（陽性検査結果が連絡され次第、ユーザーはこれを本人確認に使って即時陽性登録を実行できる）
2) 画面上に「連絡先を入力」のテキストボックスが開くので、検査を受けることが決まった時に保健所に連絡先として届けてあるメールアドレスか携帯電話番号を入力すると、陽性登録事前承認者リストに登録され所轄の保健所に届く
（保健所はこのリストに登録された連絡先の持ち主が陽性判定された連絡を検査機関から受け次第、即HER-SYSにデータを優先登録して処理番号を発行する。HER-SYSへのデータ登録完了後、連絡先は事前承認者リストから消去される）

## その他 / Additional context

1) HER-SYSシステムの運用開始当時、データ登録は保健所を通して検査を受けることが決まった段階で行われることになっており、検査結果の連絡時には登録が完了して8桁の処理番号が確定していたからこそ、検査結果連絡時に請求すれば処理番号が発行されて陽性登録を実行でき、HER-SYS処理番号を陽性登録の本人確認手段として使うことに意味があった（ただし、HER-SYSへの移行が未完了で厚労省が処理番号発行操作を代行していた自治体では、厚労省が閉庁する土日祝日に処理が停止しタイムラグが生じていた）
。

2) ところが2020年10月14日付厚労省事務連絡
https://www.mhlw.go.jp/content/000683027.pdf
でデータ登録時期が入院例を除いて「陽性結果確定時」に後ろ倒しされてしまった結果（p.7の問11「検査結果が陰性の場合も入力が必要ですか」の答え参照）、検査結果の電話連絡時点でデータ登録が完了せず処理番号が未確定という事態が常態化し、処理番号は陽性登録の本人確認用として使い物にならなくなった。

3) 2021年1月8日付厚労省事務連絡
https://www.mhlw.go.jp/content/000717197.pdf
以降は、「積極的疫学調査実施要領」及び「調査票」が改定され、検査結果連絡時に処理番号発行を請求されたら調査表の「陽性登録希望」欄の「有」をチェックして優先的にHER-SYSにデータを登録し、完了後すぐ処理番号発行操作を行うことになったものの、感染拡大時にはデータ登録が遅れることに変わりはない。

代替案1では、事前に意思表示のあった陽性登録をHER-SYSから切り離し、（発症から検査結果確定までのタイムラグに上乗せされて接触確認アプリの感染拡大防止効果の低下につながる）結果通知から陽性登録実行までのタイムラグを最小化することができる。
代替案2では、陽性登録を保健所とHER-SYSから切り離さない前提の下、事前意思表示によりデータ登録時期を少しでも前倒しして結果通知から陽性登録実行までのタイムラグを縮めることを目指している。

**Comment by keiji** [comment]
こういうケースに対応できるように、いまの接触確認APIにはPreauthorize APIが追加されていますね。

今いただいている解決策はPreauthorize APIは使わない前提に見えます。また、HER-SYSの話も絡んできそうですが、COCOAはHER-SYSと直接通信していないと理解しているので、そのあたりの整理も必要と思います。

COCOAがどう変わるのか、陽性登録サーバーやHER-SYSがどういう動きをすることになるのか。Descriptionをみる限りでは読み取れませんでした。

登録意志の事前確認というコアな部分は共感するところがあります。一方、課題に対する解決策、代替策が煮詰まっていないと感じています。

このIssueはDiscussionに移して、そこで方法について一定の方向性が出てから、あらためてIssueにするのがよいと思いますが、いかがでしょうか。

**Comment by parasite2006** [comment]
@keiji 
> こういうケースに対応できるように、いまの接触確認APIにはPreauthorize APIが追加されていますね。

ご教示ありがとうございます。そういう機能がすでに追加されているとは全く存じませんでした。
Discussionに場を移してさらに色々な角度から議論していただければ大変嬉しく存じます。
（個人的には陽性登録のプロセスを保健所とHER-SYSから切り離すべしと考えています）

**Comment by keiji** [comment]
ひとまずPreauthorize APIについてわかっていることをまとめておきます。

 * 通常のTEK取得では、実行時にシステムが確認ダイアログを出す。ユーザー操作による承認が必要。
 * Preauthorize APIを実行すると、その時点でシステムが確認ダイアログを表示する。あらかじめ利用者が承認をすると、承認から5日間、確認ダイアログなしでTEKを取得できる

Preauthorize APIを使うイメージとしては次のように考えています。

 * 検査の前に登録意思を確認してPreauthorizeすると、COCOAがいつでもTEKを取得できるようになる
 * 陽性登録サーバーから「一時キー」のようなものを発行してCOCOAに保存
 * 以後5日間、n時間ごとに一時キーを使って検査結果を問い合わせる（どこに？　陽性登録サーバー？）
 * 問い合わせた結果、陽性であった場合、陽性登録用の処理番号が返される。COCOAは端末内のTEKを取得して、感染性があった期間のTEKを陽性登録サーバーに登録する。

課題なのは、陽性登録サーバーに検査結果が登録されるわけではなく、おそらくその先にいるHER-SYSに登録されるであろうと言うことです。つまりCOCOAから陽性登録サーバーに依頼した「一時キー」の発行は、依頼を受けた陽性登録サーバーからさらにHER-SYSに転送する必要があります。
このあたりはプライバシー保護の観点から、陽性者の情報をCOCOAがアクセスするサーバーに置かない・渡さないという仕組みだと理解しています。現在も処理番号はCOCOAは関与せず、SMSなどで送信されていると理解しています。

Preauthorize APIは使えれば便利なのですが、手を入れなければならない範囲がCOCOA以外にも波及するため、慎重に検討しなければならない課題の1つと考えています。

**Reply by parasite2006** [reply]
@keiji 
お返事ありがとうございます。
読ませていただいた限りでは、どうもPreauthorization APIの機能の使用は接触確認アプリから検査結果の問い合わせができること、すなわちドイツの接触確認アプリの陽性登録システム
https://github.com/corona-warn-app/cwa-verification-server/blob/master/docs/architecture-overview.md
のように陽性登録のネットワークに検査機関が組み込まれていることが前提となっているように見受けられますね。ドイツ流の陽性登録システムには、陽性登録者のTEKを保存するサーバー（下図の赤枠）とは別に
A) 検査機関が検査結果をアップロードするTest Result Server
B) 接触確認アプリユーザーの検査結果問い合わせを仲介し、陽性登録用の処理番号（TAN）
を発行するVerification Server
の2つのサーバーが存在します。
<img width="1081" alt="GermanContactTracingSystem+" src="https://user-images.githubusercontent.com/84975567/124101632-879d8c00-da9a-11eb-84b3-8ded99970336.png">
接触確認アプリ（図の青枠）のユーザーは検査を受ける時に医師または検査機関で紙に印刷したバーコードを渡され、これをスマホでスキャンしてVerification Serverに送ると折り返し登録コード（Registaration token）が発行されます。登録コードを使って接触確認アプリ上からVerification Serverに検査結果を問い合わせ、結果が陽性なら即登録コードをVerification Serverに送って陽性登録用の処理番号（TAN）を請求できるので、検査結果確認から陽性登録実行までのタイムラグを限りなくゼロに近づけることができています。（Verification Server側は検査結果の問い合わせを受けると、送られてきた登録コードに対応するバーコードのハッシュ値をTest Result Serverに送信し、Test Result Serverは検査機関からバーコードのハッシュ値付きでアップロードされた検査結果の中から一致するハッシュ値を持つものを探し出して返信。個人である接触確認アプリユーザーが医療情報を集積したTest Result Serverに直接アクセスすることにより生じる問題をVerification Serverを介在させることにより回避）

日本の接触確認アプリの陽性登録システムには検査機関が組み込まれていないため、ユーザーは保健所の電話連絡が来るまで待たないと自分の検査結果を知ることができず、Preauthentification APIを取り入れることそのものが難しそう。

**Reply by kvaluation** [reply]
情報ありがとうございます。このスレッド内で書き込ませて下さい。

登録意思の事前確認済の際、HER-SYSに陽性登録される前に、医師による陽性診断の確定時点で、病院独自のQRコード（日次でも）を読み取るなどの操作で、すべて自動化できると一番早いでしょうか。

EN自体、Confirmed testやConfirmed clinical diagnosisをリスク評価のパラメーターとして用意しているので、陽性登録のルートを1つにしなければならないということは、ないのかなと思います(V1: TransmissionRisk, V2: Report Type)。

病院で（PCR検査又は診断のみで）陽性となったら、病院で陽性登録用のQRコードをCOCOAで読み込んでもらい、
陽性登録サーバーからの一時キーに紐付けてCOCOA側から陽性登録サーバーに陽性を登録し、
病院の陽性診断QRからのコードの正当性を自動で検証し、処理番号発行、
COCOAは一時キーに紐付いた処理番号を受信すると、TEKをアップロード。
陽性診断から数分で陽性登録が完了しそうです。

事前意思確認のない方が病院の陽性診断QRを読み取ったら、その時点で登録意思の確認ダイアログを表示するなど。

現状のHER-SYSルートはそのままで、Preauthorize APIを取り込み、陽性登録率の向上に寄与しそうな気がします。病院には多少手間ですが、接触履歴の聞き取りが楽になり、地域のクラスターの指数関数的増加を早めに抑えることができるメリットを感じていただけそうに思います。

まだまだラフなので、プライバシー保護できて、楽に早く陽性登録してもらえるような、より良い案に繋がって行くと嬉しいです。

**Reply by tmurakami** [reply]
@parasite2006

EN の Preauthorization はタイムラグ解消という観点では劇的な効果がない？のかもしれませんが、COCOA 利用者の陽性登録率向上という観点では効果があると思っています。
検査時に事前同意してさえいれば、あとは良きに計らってくれるというのは楽です。

**Comment by tmurakami** [comment]
「事前の意思表示」＝「検査結果が出る前に処理番号を取得」とした上で、

- 保健所を介さずに処理番号を発行できるように
- 陽性確定前であっても処理番号を取得できるように

というところで (実現可能性は置いておいて) ちょっと考えてみました。

HER-SYS側に以下のテーブルを用意します。

処理番号テーブル
|処理番号|電話番号|作成日時|
|-|-|-|

発症日テーブル
|電話番号|発症日トークン|発症日|トークン作成日時|
|-|-|-|-|

※ 説明のため上記構成にしていますが、もちろん発生届ID？等で結合したりしても良いと思います。

以下に流れを説明します。

1. COCOA 利用者が処理番号の発行を依頼する

電話番号を入力して、処理番号を発行します。
結果、処理番号テーブルは

|処理番号|電話番号|作成日時|
|-|-|-|
|xxxxxxxx|080xxxxxxxx|2021-07-01 10:00:00|

のようになります。

プライバシーの関係上 COCOA で電話番号を入力させるのは難しいだろうと思うので、別途専用のシステムを用意する想定です。
Web だと他人の電話番号を入力されてしまう可能性があるので、自動音声応答システム (IVR) が使えると良いのかなと考えています (本当に使えるかはわかりません)。
意味なく国のデータベースに自分の電話番号を残したい人はいないでしょうから、イタズラ電話は少ないだろうと思います。

2. COCOA 利用者が COCOA に処理番号を入力する

処理番号を入力すると、COCOA サーバー経由で HER-SYS へ発症日トークン生成リクエストを出します。
HER-SYS は発症日トークンを生成して COCOA に返します。
その際、使われた処理番号はクリア(レコード削除)されます。

結果、発症日テーブルは

|電話番号|発症日トークン|発症日|トークン作成日時|
|-|-|-|-|
|080xxxxxxxx|xxx...||2021-07-01 10:01:00|

のようになります。

発症日トークン取得後すぐに、COCOA はそのトークンを用いて、COCOA サーバー経由で HER-SYS へ発症日の取得を行います。

発症日が取得できれば COCOA はプッシュ通知を出します。
発症日が取得できなければ何もしません。

その後、一日数回程度バックグラウンドで発症日の取得を行います。

発症日トークンの有効期限は 5 日程度を想定しています。

3. HER-SYS 利用者が発生届を HER-SYS に入力する

発生届を入力すると、HER-SYS は電話番号を条件にして発症日テーブルの発症日を発生届の 発症日で更新します。
※ 発生届に発症日が記入されていない場合は検査日で代替
結果、発症日テーブルは

|電話番号|発症日トークン|発症日|トークン作成日時|
|-|-|-|-|
|080xxxxxxxx|xxx...|2021-07-01|2021-07-01 10:01:00|

のようになります。

レコードの生存期間は発症日から10日 (感染可能期間が確かこれくらい) くらいまでを想定しています。
ですので、発症日テーブルは

発症日がない場合

- トークン作成日時 + 5 日後にレコード削除

発症日がある場合

- 発症日+10日後にレコード削除
- トークン及びトークン作成日時はトークン作成日時 + 5 日後にクリア (null とか空文字とか埋める)

という感じになります。

4. COCOA 利用者が陽性登録する

発症日設定後に COCOA から (COCOA サーバー経由で) 発症日取得のリクエストがあれば HER-SYS は発症日を返します。
経路上の COCOA サーバーは陽性登録トークンを生成して、発症日と共に COCOA に返します。
結果を受けた COCOA は陽性登録を促すプッシュ通知を出します。
COCOA 利用者はプッシュ通知をタップして陽性登録画面を開き、同意の上で COCOA サーバーヘ陽性登録を行います。
登録後、不要になったトークン等を COCOA から削除します。

HER-SYS から発症日を取得しているので、COCOA 利用者に発症日を入力させずに陽性登録できるのではないかと思います。

陽性登録トークンの有効期限は 24 時間程度を想定しています。

### その他

#### 処理番号未発行の状態で発生届を入力した場合

以下のようなレコードができます。

|電話番号|発症日トークン|発症日|トークン作成日時|
|-|-|-|-|
|080xxxxxxxx||2021-07-01||

HER-SYS 側では誰が COCOA を利用しているか判断できないと思うので、発生届が入力されたら無条件で上記のレコードを作る想定です。

#### 処理番号・発症日トークンの有効期限が切れた場合

もう一度 COCOA 利用者にて処理番号発行を依頼してもらう想定です。

#### なんらかの理由で陽性登録に失敗した場合

もう一度 COCOA 利用者にて処理番号発行を依頼してもらう想定です。

発症日テーブルには以下のレコード

|電話番号|発症日トークン|発症日|トークン作成日時|
|-|-|-|-|
|080xxxxxxxx|xxx...|2021-07-01|2021-07-01 10:01:00|

が残っているはずなので、発症日トークン生成時に発症日トークンとトークン作成日時のみを上書きする想定です。
先に述べたとおり、発症日が入っているレコードは 10 日間残りますので、その期間であれば COCOA 利用者は (処理番号発行すれば) 何度でも陽性登録できることになります。

#### 検査結果が陰性だった場合

検査結果が陰性だったときは COCOA 利用者に何の通知もされません。
検査結果は別途電話等で連絡されるでしょうから、別にいいかな？と思っています。

---

と長々と書いてみましたが、いろいろと大変そうなので (私の案はどうあれ) やるにしても時間がかかるだろうなという印象です。

私見ですが、タイムラグがあるとすれば、原因は発生届の代行入力なのではないかな？と勝手に思っています。
検査を受けつけている病院側で HER-SYS に発生届を入力できていれば、保健所側は陽性者の登録意思を確認後に「処理番号発行」ボタンを押すだけで良いので、それほど遅延するようには思えないのです。

もちろん私の想像していない別の何かが原因ということも十分ありえるのですが、私の推測が正しいのであれば、事前の意思表示は (保健所の負荷は下がるとは思いますが) タイムラグの解消という観点では効果が薄いのではないかと思っています。

- 相当のタイムラグとは何日程度なのか
- 原因はなんなのか

について、COCOA 利用者や各自治体等へアンケートとかできると良いのかなと思います。
※ 既に何かしらの調査結果が公開されているのであればすみません。

あと、HER-SYS から COCOA に関連する部分を別システムにするという案がありましたが、個人的には厳しいのではないかと思っています。
私の案を元にすると、そのシステムに必要な入力項目は「電話番号」「発症日」くらいだと思うので一見簡単そうですが、その他にもログイン機能や操作履歴閲覧機能等が必要になる気がしますし、どこに配置するかといったこと等も検討が必要で、構築自体ちょっと大変そうです。

また、これも私の案を元にお話ししますが、私の案では医療機関であれ保健所であれ HER-SYS に発生届を登録したタイミングで陽性登録できるようになっているのですが、COCOA 関連を別システムにしてしまうと、少なくとも医療機関は情報登録しなくなってしまうだろうと思っています。
なぜなら COCOA は積極的疫学調査の補助ツールであり、医師の業務とは関係がないので、そのシステムを導入する動機がないからです。
ですので、そのシステムを使うのは保健所とか感染研とか積極的疫学調査を業務とする機関になると思うのですが、HER-SYS とは別にもうひとつシステムを使うというのも、なかなか大変なのではないかと懸念しています。
いくら操作が単純であったとしても、システムを切り替えながら作業をするのは精神的に疲れてしまうのではないかと思っています。

深く考えた内容ではないので、失礼だったり考慮不足な文章になっているかもしれません。
ちょっとまとまりのない内容になっているかもしれませんが、ご容赦下さい。

**Reply by parasite2006** [reply]
@tmurakami 

> 私見ですが、タイムラグがあるとすれば、原因は発生届の代行入力なのではないかな？と勝手に思っています。

医療機関ではHER-SYSへのデータ入力につけられる人手がないので発生届を従来通り紙に書いて保健所にFAXで送ってしまい、その結果ちょっと感染拡大すると保健所では発生届のFAXが溜まり、人手に頼るデータ入力が追いつかなくなります。しかも8桁の処理番号はHER-SYSへのデータ登録が完了しないと確定しないので、去年の秋から始まった第3波の前半、特に年が変わる前には「宿泊療養に入るときに処理番号を請求したのに、届いたのは2週間の健康観察期間が終わる直前だった」という話をTwitter上で複数回聞いたことがあります。

2021年1月8日付厚労省事務連絡
https://www.mhlw.go.jp/content/000717197.pdf
以降は、「積極的疫学調査実施要領」及び「調査票」が改定され、検査結果連絡時に処理番号発行を請求されたら調査表の「陽性登録希望」欄の「有」をチェックして優先的にHER-SYSにデータを登録し、完了後すぐ処理番号発行操作を行うことになったため、請求した処理番号が届くまでのタイムラグは2-3日程度まで縮まっていますが、県庁所在地か政令指定都市単位でHER-SYSデータ登録センターでも設けて医療機関からの発生届のFAXをここに集め、FAXが届く端からOCRでデジタル化して自動的にHER-SYS登録用のテンプレートに流し込むぐらいのことをしないと、感染拡大のたびに人手によるデータ登録の負担が保健所に集中して登録遅れが繰り返されるのは避けられないだろうと考えています。

厚労省HPのHER-SYS特設コーナー
https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/0000121431_00129.html
にリンクされている沖縄県での活用例
https://www.mhlw.go.jp/content/10900000/000687829.pdf
を除いて、医療機関からのHER-SYSデータ登録はまだまだ進んでいないと思われます。厚労省は接触確認アプリの活用状況を都道府県別の処理番号発行件数と陽性登録件数で見た調査結果（2021年6月18日開催の厚労省の第39回新型コロナウイルス感染症対策アドバイザリーボード配布資料）
https://www.mhlw.go.jp/content/10900000/000793730.pdf
をまとめることができるなら、それとあわせて接触確認アプリの陽性登録の効率を左右する（しかも現状ではその足を引っ張っている）HER-SYSのデータ登録の状況（医療機関からのデータ登録件数と陽性結果判明からHER-SYS登録完了までの所要日数）も都道府県別にまとめてもらいたいものです。

コードとは明らかに離れた話になってしまってどうもすみませんでした。これで終わりにします。

**Reply by tmurakami** [reply]
> これで終わりにします。

追記することをご容赦下さい。

EN 仕様を見る限り「陽性者」＝「検査で陽性判定された者 (つまり検査精度を考慮しない) 」と定義しているように感じています。
一方、我が国においては「陽性者」＝「(検査の結果等から) 医師によって陽性と判断された者」と定義しているように感じています。
※ 我が国においては、例えば複数回検査によって陰性の割合が多ければ発生届が提出されない可能性もあるのではないかと思っています。
根本的には、このズレが遅延を生んでいるように感じます。

例示頂いたドイツにおいては、EN に準じて検査機関が「検査で陽性判定された者 (つまり検査精度を考慮しない) 」をシステムに登録している？ようですが、我が国の「陽性者」の定義が「(検査の結果等から) 医師によって陽性と判断された者」ということであるならば、検査機関が介入する余地はないように思います。
※ EN 仕様と COCOA とで少し認識が異なるのかもしれないと感じたのは https://github.com/cocoa-mhlw/cocoa/issues/26#issuecomment-790340112 からです。

COCOA 的に「陽性者」＝「検査で陽性判定された者 (つまり検査精度を考慮しない) 」とするのであれば、個人的には HER-SYS と独立した陽性登録システムがあって良いと思っています。
そのシステムは検査機関でしか使用されないでしょうし、そのほうがシンプルであるように思っています。
ですがその場合は、日々発表される「感染者」とCOCOA の「陽性者」は定義が異なることになるでしょうし、また「偽陽性」であっても陽性登録できることになります。

一方、COCOA 的に「陽性者」＝「(検査の結果等から) 医師によって陽性と判断された者」とするならば、定義としては「陽性者」＝日々発表される「感染者」となりますが、HER-SYS に発生届が入力されない限り陽性者には成り得ないことになると思うので、(発生届代行がタイムラグの原因であるならば) TEK の登録までにタイムラグが発生することになるかと思います。

**Comment by tmurakami** [comment]
@keiji

EN API の Preauthorization について聞きたいことがあります。
お暇なときにでも回答頂ければ。

- 同意取り消しはできますか？

同意の取り消しは API として提供されていない気がしていて、取り消しできるようにするためには COCOA 側でフラグを持つ等して作り込む必要があると思っていますが、合っていますか？

- 再同意はできますか？

例えば二度事前同意をする (Android では ` requestPreAuthorizedTemporaryExposureKeyHistory()`を、iOS では`preAuthorizeDiagnosisKeys(completionHandler:)`を二回呼び出す)場合、二度目はダイアログが表示されますでしょうか？
同意取り消しを COCOA 側で作り込んだとして、EN API と不整合が生じるのかどうかが気になっています。

**Reply by keiji** [reply]
> 同意取り消しはできますか？
> 同意の取り消しは API として提供されていない気がしていて、取り消しできるようにするためには COCOA 側でフラグを持つ等して作り込む必要があると思っていますが、合っていますか？

EN APIには同意の取り消し機能はありません。取り消しについては、COCOA側で処理する必要があります。

> 例えば二度事前同意をする (Android では requestPreAuthorizedTemporaryExposureKeyHistory()を、iOS ではpreAuthorizeDiagnosisKeys(completionHandler:)を二回呼び出す)場合、二度目はダイアログが表示されますでしょうか？

いいえ。
`requestPreAuthorizedTemporaryExposureKeyHistory`を複数回実行しても、表示されるダイアログは最初の一回だけです。
ダイアログで「許可しない」を選択すると、同意ダイアログがしばらく表示されない現象が確認されています。これについては正常エラー系の処理をもう一度見直してみます。

**Reply by tmurakami** [reply]
早速の回答ありがとうございます。
同意取り消しを実装するとなると、考慮すべき点がありそうですね。

**Reply by Ryanaka** [reply]
Androidでは同意状態を変更できそうなUIがありますが、こちらを操作したときの挙動はどうなるのでしょうか。
<img src="https://user-images.githubusercontent.com/9909193/123897940-4e342600-d99f-11eb-96c8-9fc0fb6beab4.png" width="75%">

**Reply by keiji** [reply]
画像のアップロードに失敗しているようです

**Reply by Ryanaka** [reply]
失礼しました。再アップロードします。

**Reply by keiji** [reply]
ありがとうございます。確認しました。
許可の取り消しはAPIではサポートしていませんが、ご指摘の通り、設定の「接触通知システム」のアプリ詳細から開いた画面から許可を取り消せることがわかりました。

https://user-images.githubusercontent.com/932136/123900883-e1bc2580-d9a4-11eb-850f-499eef49522a.mp4

この設定は通常では無効化されていますが、一度アプリからPreauthorizeを許可すると、ユーザーが自由にON/OFFできます（再許諾や取り消し）。その後、Relase（5日以内に1回だけ許されるTEKの読み取り）をした後は、ふたたび無効化されて、アプリからPreauthorizeするまでユーザーが触ることができなくなります。

iOSにも設定の「接触確認アプリの項目を開いた中にある「接触通知」に同様の設定があることが確認できました。

![Screen Shot 2021-06-30 at 13 18 29](https://user-images.githubusercontent.com/932136/123901380-c867a900-d9a5-11eb-8380-cad047b5a674.png)

**Comment by keiji** [comment]
ちなみにCOCOAの実装の話とは外れますが、陽性登録のサポートと言う文脈ではこんな情報もあります。

https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/cocoa_00138.html#h2_free5

これまで保健所からしかできなかった処理番号の発行を、利用者がMy HER-SYSや自動架電（電話がかかってきて健康観察を入力するシステム）から発行できるようになっています（７月１日から）。

**Reply by parasite2006** [reply]
自宅/宿泊療養者が健康観察データをHER-SYSに自己入力するためのMY HER-SYS IDも、COCOAの陽性登録用処理番号と同様HER-SYSへのデータ登録が完了しないと確定しないので、検査結果の電話連絡を受けてからMY HER-SYS IDを渡されるまでのタイムラグは残りますが、検査結果の電話連絡時に口頭で処理番号を請求し損ねてもMY HER-SYS IDがあれば自力で請求できるようになったことは大きな前進ですね。接触確認アプリの特設コーナーに案内チラシの画像が大きく出ているのもナイスで
https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/cocoa_00138.html#h2_free5
この情報をCOCOA内での処理番号取得方法の説明でも表示して周知していただければありがたいです。

**Reply by keiji** [reply]
My HER-SYSについて、実はちょろっと書いていたりします。

https://github.com/cocoa-mhlw/cocoa/discussions/253

ひとまずこの画面でいって、もしわかりにくいという意見があれば、説明や表現について見直していきたいですね。

**Reply by parasite2006** [reply]
#253 にご提示いただいている説明画像、やはりこちら
https://github.com/cocoa-mhlw/cocoa/issues/201#issuecomment-867383028
で提案させていただいている通り、検査を受けることが決まってから結果の電話連絡が来るまでのできるだけ早い段階であらかじめユーザーに見ておいてもらえるようホーム画面上にもリンクを置いておいていただけたらと存じます。

**Reply by keiji** [reply]
ありがとうございます。いただいた提案について開発チームに共有します。

また今回My HER-SYSや自動架電での処理番号発行が可能になった実態と合わせるため #201 でいただいているテキストの調整も別の意味で課題になりますね。テキストを追加するだけでは利用者が読まなくなるだけなので、必要な情報を過不足なく入れることが課題になりそうです。
